eth-feehistory docs (#4190)
ChainSafe/web3.js: docs/web3-eth.rst

------------------------------------------------------------------------------

.. _eth-feehistory:

getFeeHistory
=====================

.. code-block:: javascript

    web3.eth.getFeeHistory(blockCount, newestBlock, rewardPercentiles, [callback])

Transaction fee history
Returns base fee per gas and transaction effective priority fee per gas history for the requested block range if available. 
The range between headBlock-4 and headBlock is guaranteed to be available while retrieving data from the pending block and older
history are optional to support. For pre-EIP-1559 blocks the gas prices are returned as rewards and zeroes are returned for the base fee per gas.

----------
Parameters
----------

> 1. ``String`` - Number of blocks in the requested range. Between 1 and 1024 blocks can be requested in a single query. Less than requested may be returned if not all blocks are available.

Reviewer: ```suggestion
1. ``String|Number|BN|BigNumber`` - Number of blocks in the requested range. Between 1 and 1024 blocks can be requested in a single query. Less than requested may be returned if not all blocks are available.
```

Reviewer 2: Shouldn't `blockCount` just be a `number`? The method takes an int, https://github.com/ethereum/go-ethereum/blob/c49e065fea78a5d3759f7853a608494913e5824e/internal/ethapi/api.go#L90

I guess it depends on if an inputFormatter is applied to force it to format it to a number?

Reviewer 3: Unfortunately its a design decision we need to stick with based on the history of web3.js, the 4.x branch improves on this a bit.

For reference check out [eth_gettransactionfromblock](https://web3js.readthedocs.io/en/v1.4.0/web3-eth.html#gettransactionfromblock)

> 2. ``String`` - Highest number block of the requested range.

Reviewer: ```suggestion
2. ``String|Number|BN|BigNumber`` - Highest number block of the requested range.
```

> 3. ``Array of numbers`` - (optional) A monotonically increasing list of percentile values to sample from each block's effective priority fees per gas in ascending order, weighted by gas used.

Reviewer 3: can we maybe include an example of a `monotonically` list

> 3. ``Array of numbers`` - (optional) A monotonically increasing list of percentile values to sample from each block's effective priority fees per gas in ascending order, weighted by gas used. Example: `["0", "25", "50", "75", "100"]` or `["0", "0.5", "1", "1.5", "3", "80"]`

Reviewer: ```suggestion
3. ``Array of numbers`` - A monotonically increasing list of percentile values to sample from each block's effective priority fees per gas in ascending order, weighted by gas used. Example: `["0", "25", "50", "75", "100"]` or `["0", "0.5", "1", "1.5", "3", "80"]`
```

4. ``Function`` - (optional) Optional callback, returns an error object as first parameter and the result as second.

-------
Returns
-------

``Promise`` returns ``Object`` - Fee history for the returned block range. The object:

> - ``String`` oldestBlock  - Lowest number block of the returned range.

Reviewer: ```suggestion
- ``Number`` oldestBlock  - Lowest number block of the returned range.
```

- ``Array of strings`` baseFeePerGas  - An array of block base fees per gas. This includes the next block after the newest of the returned range, because this value can be derived from the newest block. Zeroes are returned for pre-EIP-1559 blocks.
- ``Array of numbers`` gasUsedRatio  - An array of block gas used ratios. These are calculated as the ratio of gasUsed and gasLimit.

> - ``Array of strings`` reward  - An array of effective priority fee per gas data points from a single block. All zeroes are returned if the block is empty.

Reviewer: ```suggestion
- ``Array of string arrays`` reward  - An array of effective priority fee per gas data points from a single block. All zeroes are returned if the block is empty.
```
